_posts/2021-03-15-service-provider-interface.md

```diff
@@ -0,0 +1,201 @@
+    book = library.getBook(name);
+  }
+  return book;
+}
+```
+This class will:
+
+- Invoke the static factory method `load()` to get an instance of `ServiceLoader`.
+- Search and retrieve all available implementations using the `iterate()` method.
+  The client will get a singleton `LibraryService` object to retrieve the books.
+
+#### standard-library
+First, we include the dependency to the service API provider in the pom file of this submodule.
+```xml
+<dependency>
+  <groupId>org.library</groupId>
+  <artifactId>library-service-provider</artifactId>
+  <version>1.0-SNAPSHOT</version>
+</dependency>
+```
+Then we create a class that implements the Library SPI:
+```java
+public class StandardLibrary implements Library {
+
+  private final Map<String, Book> books;
+
+  public StandardLibrary() {
+    books = new TreeMap<>();
+    Book nineteenEightyFour = new Book("Nineteen Eighty-Four", 
+            "George Orwell", "Description");
+    Book theLordOfTheRings = new Book("The Lord of the Rings", 
+            "J. R. R. Tolkien", "Description");
+
+    books.put("Nineteen Eighty-Four", nineteenEightyFour);  
+    books.put("The Lord of the Rings", theLordOfTheRings);  
+  }
+
+  @Override
+  public Book getBook(String name) {
+    return books.get(name);
+  }
+}
+```
+This implementation will give the possibility for the client to get the book's information if available.
+Finally, we should create a folder called `META-INF/services` in the resources directory with a file named `org.library.spi.Library`. This file will contain the full class name of the implementation that will be used by the `ServiceLoader` to instantiate it. In our case, il will be `org.library.StandardLibrary`.
+#### custom-library
+The custom-library submodule has the same structure and requirements as the standard-library submodule. However, the implementation of the Library SPI, the file name and the class name that will be created in the `META-INF/services` folder will change.
+#### library-demo
+In this submodule, we will call the `serviceLibrary` to get information about some books.
+In the beginning, we will use only the standard library as a library for our demo, then we will see how we can add more capabilities to our demo project by adding the custom-library jar file to the classpath. The ServideLoader will then load and instantiate our provider.
+In the first place, let's add the standard-library submodule to the library-demo `pom` file:
+```xml
+<dependency>
+  <groupId>org.library</groupId>
+  <artifactId>standard-library</artifactId>
+  <version>1.0-SNAPSHOT</version>
+</dependency>
+```
+Then, we try to get information about two books:
+```java
+public static void main(String[] args) {
+  LibraryService libraryService = LibraryService.getInstance();
+  manageBookRequest("Clean Code", libraryService);
+  manageBookRequest("The Lord of the Rings", libraryService);
+}
+
+private static void manageBookRequest(String bookName, LibraryService library) {
```

Comment:
maybe rename this method to `requestBook()`. `manageBookRequest` sounds very enterprisey :)